File: backend/tests/test_production_security.py
"""
Production Security, Hardening & Final API Validation Test Suite
Uses standard Python unittest and FastAPI TestClient with isolated dependencies.
"""

import unittest
import sys
import os
from unittest.mock import MagicMock

# Add backend directory to sys.path
backend_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
if backend_dir not in sys.path:
    sys.path.insert(0, backend_dir)

from fastapi.testclient import TestClient
from app.main import app
from app.database.connection import get_database
from app.utils.security import (
    hash_password,
    verify_password,
    create_access_token,
    decode_access_token,
    get_current_user,
    require_admin,
)


class TestProductionSecurity(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.client = TestClient(app)

    def tearDown(self):
        app.dependency_overrides.clear()

    def test_01_security_headers(self):
        """Verify OWASP security headers are present on API responses."""
        response = self.client.get("/health")
        self.assertEqual(response.status_code, 200)
        headers = response.headers
        self.assertEqual(headers.get("x-content-type-options"), "nosniff")
        self.assertEqual(headers.get("x-frame-options"), "DENY")
        self.assertIn("x-xss-protection", headers)
        self.assertIn("referrer-policy", headers)

    def test_02_health_check_safe_response(self):
        """Verify health endpoint does not expose credentials or database URI."""
        response = self.client.get("/health")
        data = response.json()
        self.assertIn("status", data)
        self.assertIn("service", data)
        raw_text = response.text.lower()
        self.assertNotIn("mongodb+srv", raw_text)
        self.assertNotIn("password", raw_text)
        self.assertNotIn("jwt_secret", raw_text)

    def test_03_password_hashing_argon2(self):
        """Verify password hashing uses secure Argon2 and verifies correctly."""
        plain = "SecureTestPassword123!"
        hashed = hash_password(plain)
        self.assertNotEqual(hashed, plain)
        self.assertIn("$argon2", hashed)
        self.assertTrue(verify_password(plain, hashed))
        self.assertFalse(verify_password("WrongPassword", hashed))

    def test_04_jwt_token_creation_and_expiration(self):
        """Verify JWT access tokens encode, sign, and decode securely."""
        payload = {"sub": "test_user_id_123", "role": "customer", "mobile": "9876543210"}
        token = create_access_token(payload)
        self.assertIsInstance(token, str)
        self.assertGreater(len(token), 20)

        decoded = decode_access_token(token)
        self.assertIsNotNone(decoded)
        self.assertEqual(decoded.get("sub"), "test_user_id_123")
        self.assertEqual(decoded.get("role"), "customer")
        self.assertIn("exp", decoded)

    def test_05_unauthenticated_request_rejected(self):
        """Verify protected endpoints return 401 Unauthorized without a valid token."""
        endpoints = [
            "/api/profile/me",
            "/api/kyc/me",
            "/api/holdings/me",
            "/api/withdrawals",
            "/api/transactions",
            "/api/notifications",
            "/api/admin/dashboard",
            "/api/admin/users",
            "/api/admin/rates",
        ]
        for ep in endpoints:
            resp = self.client.get(ep)
            self.assertEqual(resp.status_code, 401, f"Expected 401 for unauthenticated {ep}, got {resp.status_code}")

    def test_06_customer_cannot_access_admin_endpoints(self):
        """Verify customer role is strictly denied access (403 Forbidden) on Admin endpoints."""
        mock_customer = {
            "id": "67b96000e783457a4eb182a1",
            "role": "customer",
            "name": "Customer User",
            "mobile": "9876543210",
        }
        app.dependency_overrides[get_current_user] = lambda: mock_customer

        admin_endpoints = [
            "/api/admin/dashboard",
            "/api/admin/users",
            "/api/admin/kyc/pending",
            "/api/admin/rates",
            "/api/admin/purchases",
            "/api/admin/withdrawals",
            "/api/admin/transactions",
            "/api/admin/notifications",
        ]
        for ep in admin_endpoints:
            resp = self.client.get(ep)
            self.assertEqual(resp.status_code, 403, f"Expected 403 Forbidden for customer on {ep}, got {resp.status_code}")

    def test_07_rate_and_purchase_validation(self):
        """Verify purchase payload rejects negative, zero, or invalid quantities."""
        mock_customer = {"id": "67b96000e783457a4eb182a1", "role": "customer"}
        app.dependency_overrides[get_current_user] = lambda: mock_customer
        app.dependency_overrides[get_database] = lambda: MagicMock()

        # Negative quantity
        resp = self.client.post("/api/purchases", json={"metal": "gold", "quantity_grams": -1.5})
        self.assertEqual(resp.status_code, 422)

        # Zero quantity
        resp = self.client.post("/api/purchases", json={"metal": "gold", "quantity_grams": 0})
        self.assertEqual(resp.status_code, 422)

        # Invalid metal
        resp = self.client.post("/api/purchases", json={"metal": "platinum", "quantity_grams": 1.0})
        self.assertEqual(resp.status_code, 422)

    def test_08_withdrawal_validation_rejects_invalid_inputs(self):
        """Verify withdrawal payload rejects negative or invalid parameters."""
        mock_customer = {"id": "67b96000e783457a4eb182a1", "role": "customer"}
        app.dependency_overrides[get_current_user] = lambda: mock_customer
        app.dependency_overrides[get_database] = lambda: MagicMock()

        # Negative quantity
        resp = self.client.post("/api/withdrawals", json={"metal": "gold", "quantity_grams": -0.5})
        self.assertEqual(resp.status_code, 422)

        # Invalid metal
        resp = self.client.post("/api/withdrawals", json={"metal": "diamond", "quantity_grams": 1.0})
        self.assertEqual(resp.status_code, 422)

    def test_09_kyc_validation(self):
        """Verify KYC payload validates required fields, gender, and ID document types."""
        mock_customer = {"id": "67b96000e783457a4eb182a1", "role": "customer"}
        app.dependency_overrides[get_current_user] = lambda: mock_customer
        app.dependency_overrides[get_database] = lambda: MagicMock()

        # Invalid gender
        invalid_gender_payload = {
            "full_name": "Test User",
            "date_of_birth": "1990-01-01",
            "gender": "unknown_gender",
            "address": {
                "address_line": "123 Main St",
                "city": "Chennai",
                "state": "Tamil Nadu",
                "pincode": "600001",
            },
            "id_type": "pan",
            "id_number": "ABCDE1234F",
        }
        resp = self.client.post("/api/kyc/submit", json=invalid_gender_payload)
        self.assertEqual(resp.status_code, 422)

    def test_10_send_otp_new_user(self):
        """Verify send OTP endpoint sends dev OTP for an un-registered mobile number."""
        mock_db = MagicMock()
        mock_db.users.find_one.return_value = None  # Not registered
        app.dependency_overrides[get_database] = lambda: mock_db

        resp = self.client.post("/api/auth/send-otp", json={"mobile": "9123456780", "purpose": "signup"})
        self.assertEqual(resp.status_code, 200)
        data = resp.json()
        self.assertTrue(data.get("otp_sent"))
        self.assertEqual(data.get("dev_otp"), "123456")

    def test_11_send_otp_existing_user_rejected(self):
        """Verify send OTP for signup rejects an already registered mobile number."""
        mock_db = MagicMock()
        mock_db.users.find_one.return_value = {"mobile": "9876543210", "role": "customer"}
        app.dependency_overrides[get_database] = lambda: mock_db

        resp = self.client.post("/api/auth/send-otp", json={"mobile": "9876543210", "purpose": "signup"})
        self.assertEqual(resp.status_code, 400)
        self.assertIn("already registered", resp.json().get("detail", ""))

    def test_12_verify_otp_creates_user_and_token(self):
        """Verify valid OTP creates new user account and returns JWT token with profile_completed: False."""
        mock_db = MagicMock()
        mock_db.users.find_one.return_value = None
        mock_db.users.insert_one.return_value = MagicMock(inserted_id="67b96000e783457a4eb182a1")
        app.dependency_overrides[get_database] = lambda: mock_db

        resp = self.client.post(
            "/api/auth/verify-otp",
            json={"mobile": "9123456780", "otp": "123456", "name": "Ravi Kumar", "purpose": "signup"}
        )
        self.assertEqual(resp.status_code, 200)
        data = resp.json()
        self.assertIn("access_token", data)
        self.assertEqual(data.get("user", {}).get("profile_completed"), False)

    def test_13_verify_otp_invalid_rejected(self):
        """Verify wrong OTP code is rejected with 400 Bad Request."""
        mock_db = MagicMock()
        mock_db.otps.find_one.return_value = None
        app.dependency_overrides[get_database] = lambda: mock_db

        resp = self.client.post(
            "/api/auth/verify-otp",
            json={"mobile": "9123456780", "otp": "999999", "name": "Ravi Kumar", "purpose": "signup"}
        )
        self.assertEqual(resp.status_code, 400)
    def test_14_direct_registration_creates_account_and_jwt_token(self):
        """Verify direct user registration hashes password, creates account, and returns JWT token."""
        mock_db = MagicMock()
        mock_db.users.find_one.return_value = None
        mock_db.users.insert_one.return_value = MagicMock(inserted_id="67b96000e783457a4eb182b2")
        app.dependency_overrides[get_database] = lambda: mock_db

        resp = self.client.post(
            "/api/auth/register",
            json={
                "name": "Karthik Raja",
                "mobile": "9845123456",
                "email": "karthik@example.com",
                "password": "StrongPassword123!",
            }
        )
        self.assertEqual(resp.status_code, 201)
        data = resp.json()
        self.assertIn("access_token", data)
        self.assertEqual(data["user"]["name"], "Karthik Raja")
        self.assertEqual(data["user"]["profile_completed"], False)

    def test_15_duplicate_registration_rejected(self):
        """Verify registration rejects duplicate mobile numbers."""
        mock_db = MagicMock()
        mock_db.users.find_one.return_value = {"mobile": "9845123456", "role": "customer"}
        app.dependency_overrides[get_database] = lambda: mock_db

        resp = self.client.post(
            "/api/auth/register",
            json={
                "name": "Karthik Raja",
                "mobile": "9845123456",
                "email": "karthik@example.com",
                "password": "StrongPassword123!",
            }
        )
        self.assertEqual(resp.status_code, 400)
        self.assertIn("already registered", resp.json().get("detail", ""))


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: backend/tests/testProductionSecurity.js
import assert from 'node:assert';
import http from 'node:http';
import axios from 'axios';
import { createApp } from '../src/app.js';
import { initDatabase, query } from '../src/config/db.js';
import { hashPassword, verifyPassword, createAccessToken, verifyAccessToken } from '../src/utils/security.js';

async function runTests() {
  console.log('====================================================');
  console.log('Running Production Security & API Validation Tests...');
  console.log('====================================================');

  await initDatabase();
  const app = createApp();

  const server = http.createServer(app);
  await new Promise((resolve) => server.listen(0, '127.0.0.1', resolve));
  const port = server.address().port;
  const baseUrl = `http://127.0.0.1:${port}`;
  const client = axios.create({ baseURL: baseUrl, validateStatus: () => true });

  let passed = 0;
  let failed = 0;

  async function test(name, fn) {
    try {
      await fn();
      console.log(`  [PASS] ${name}`);
      passed++;
    } catch (err) {
      console.error(`  [FAIL] ${name}:`, err.message);
      failed++;
    }
  }

  // 1. Security Headers
  await test('test_01_security_headers: OWASP headers present', async () => {
    const res = await client.get('/health');
    assert.strictEqual(res.status, 200);
    assert.strictEqual(res.headers['x-content-type-options'], 'nosniff');
    assert.strictEqual(res.headers['x-frame-options'], 'SAMEORIGIN');
  });

  // 2. Health check safe response
  await test('test_02_health_check_safe_response: does not expose credentials', async () => {
    const res = await client.get('/health');
    assert.strictEqual(res.status, 200);
    assert.strictEqual(res.data.status, 'healthy');
    const raw = JSON.stringify(res.data).toLowerCase();
    assert.strictEqual(raw.includes('password'), false);
    assert.strictEqual(raw.includes('jwt_secret'), false);
    assert.strictEqual(raw.includes('mysql://'), false);
  });

  // 3. Password Hashing
  await test('test_03_password_hashing: secure bcrypt hashing and verification', async () => {
    const plain = 'SecureTestPassword123!';
    const hashed = await hashPassword(plain);
    assert.notStrictEqual(hashed, plain);
    assert.strictEqual(await verifyPassword(plain, hashed), true);
    assert.strictEqual(await verifyPassword('WrongPassword', hashed), false);
  });

  // 4. JWT Token Creation and Expiration
  await test('test_04_jwt_token_creation_and_expiration: sign and verify JWT', async () => {
    const payload = { sub: 'test_user_id_123', role: 'customer', mobile: '9876543210' };
    const token = createAccessToken(payload);
    assert.ok(typeof token === 'string' && token.length > 20);

    const decoded = verifyAccessToken(token);
    assert.ok(decoded);
    assert.strictEqual(decoded.sub, 'test_user_id_123');
    assert.strictEqual(decoded.role, 'customer');
    assert.ok(decoded.exp);
  });

  // 5. Unauthenticated requests rejected (401)
  await test('test_05_unauthenticated_request_rejected: 401 on protected routes', async () => {
    const endpoints = [
      '/api/profile/me',
      '/api/kyc/me',
      '/api/holdings/me',
      '/api/withdrawals',
      '/api/transactions',
      '/api/notifications',
      '/api/admin/dashboard',
      '/api/admin/users',
      '/api/admin/rates',
    ];
    for (const ep of endpoints) {
      const res = await client.get(ep);
      assert.strictEqual(res.status, 401, `Expected 401 for ${ep}, got ${res.status}`);
    }
  });

  // Create a customer user in DB for customer role tests
  const testCustomerMobile = '9845000001';
  await query('DELETE FROM users WHERE mobile = ?', [testCustomerMobile]);
  const regRes = await client.post('/api/auth/register', {
    name: 'Test Customer',
    mobile: testCustomerMobile,
    email: 'testcustomer@example.com',
    password: 'TestPassword123!',
  });
  const customerToken = regRes.data.access_token;
  const customerAuth = { headers: { Authorization: `Bearer ${customerToken}` } };

  // 6. Customer cannot access Admin endpoints (403)
  await test('test_06_customer_cannot_access_admin_endpoints: 403 Forbidden for customer', async () => {
    const adminEndpoints = [
      '/api/admin/dashboard',
      '/api/admin/users',
      '/api/admin/kyc/pending',
      '/api/admin/rates',
      '/api/admin/purchases',
      '/api/admin/withdrawals',
      '/api/admin/transactions',
      '/api/admin/notifications',
    ];
    for (const ep of adminEndpoints) {
      const res = await client.get(ep, customerAuth);
      assert.strictEqual(res.status, 403, `Expected 403 for customer on ${ep}, got ${res.status}`);
    }
  });

  // 7. Rate and Purchase Validation
  await test('test_07_rate_and_purchase_validation: rejects invalid metals and quantities', async () => {
    // Negative quantity
    const negRes = await client.post('/api/purchases', { metal: 'gold', quantity_grams: -1.5 }, customerAuth);
    assert.strictEqual(negRes.status, 400);

    // Invalid metal
    const metalRes = await client.post('/api/purchases', { metal: 'platinum', quantity_grams: 1.0 }, customerAuth);
    assert.strictEqual(metalRes.status, 400);
  });

  // 8. Withdrawal Validation
  await test('test_08_withdrawal_validation_rejects_invalid_inputs: rejects invalid inputs', async () => {
    // KYC not verified yet -> 403
    const res = await client.post('/api/withdrawals', { metal: 'gold', quantity_grams: 1.0 }, customerAuth);
    assert.strictEqual(res.status, 403);
  });

  // 9. KYC Submission
  await test('test_09_kyc_submission: submits KYC documents successfully', async () => {
    const kycPayload = {
      full_name: 'Test Customer Full',
      date_of_birth: '1995-05-15',
      gender: 'Male',
      address: {
        address_line: '123 Test Street',
        city: 'Chennai',
        state: 'Tamil Nadu',
        pincode: '600001',
      },
      id_type: 'Aadhaar',
      id_number: '123456789012',
    };
    const res = await client.post('/api/kyc/submit', kycPayload, customerAuth);
    assert.strictEqual(res.status, 201);
    assert.strictEqual(res.data.status, 'pending');
  });

  // 10. Direct Registration
  await test('test_10_direct_registration_creates_account_and_jwt_token: registration', async () => {
    const mobile = '9845000002';
    await query('DELETE FROM users WHERE mobile = ?', [mobile]);
    const res = await client.post('/api/auth/register', {
      name: 'Karthik Raja',
      mobile,
      email: 'karthik@example.com',
      password: 'StrongPassword123!',
    });
    assert.strictEqual(res.status, 201);
    assert.ok(res.data.access_token);
    assert.strictEqual(res.data.user.name, 'Karthik Raja');
    assert.strictEqual(res.data.user.profile_completed, false);
  });

  // 11. Duplicate Registration Rejected
  await test('test_11_duplicate_registration_rejected: rejects duplicate mobile', async () => {
    const res = await client.post('/api/auth/register', {
      name: 'Karthik Raja Duplicate',
      mobile: '9845000002',
      email: 'karthik2@example.com',
      password: 'StrongPassword123!',
    });
    assert.strictEqual(res.status, 400);
    assert.ok(res.data.detail.includes('already registered'));
  });

  // 12. Login with valid credentials
  await test('test_12_login_with_valid_credentials: login success', async () => {
    const res = await client.post('/api/auth/login', {
      mobile: '9845000002',
      password: 'StrongPassword123!',
    });
    assert.strictEqual(res.status, 200);
    assert.ok(res.data.access_token);
    assert.strictEqual(res.data.user.mobile, '9845000002');
  });

  // 13. Login with invalid password rejected
  await test('test_13_login_with_invalid_password_rejected: rejects wrong password', async () => {
    const res = await client.post('/api/auth/login', {
      mobile: '9845000002',
      password: 'WrongPassword123!',
    });
    assert.strictEqual(res.status, 401);
    assert.ok(res.data.detail.includes('Invalid'));
  });

  // 14. Send OTP and verify OTP
  await test('test_14_send_otp_and_verify_otp: OTP dispatch and dev verify', async () => {
    const sendRes = await client.post('/api/auth/send-otp', { mobile: '+919876543210' });
    assert.strictEqual(sendRes.status, 200);
    assert.strictEqual(sendRes.data.mobile, '9876543210');

    const verifyRes = await client.post('/api/auth/verify-otp', { mobile: '9876543210', otp: '123456' });
    assert.strictEqual(verifyRes.status, 200);
    assert.strictEqual(verifyRes.data.valid, true);
  });

  // 15. Verify OTP invalid rejected
  await test('test_15_verify_otp_invalid_rejected: rejects invalid OTP', async () => {
    const res = await client.post('/api/auth/verify-otp', { mobile: '9876543210', otp: '000000' });
    assert.strictEqual(res.status, 400);
    assert.ok(res.data.detail.includes('Invalid or expired OTP'));
  });

  server.close();

  console.log('====================================================');
  console.log(`Results: ${passed} passed, ${failed} failed.`);
  console.log('====================================================');

  if (failed > 0) {
    process.exit(1);
  }
  process.exit(0);
}

runTests().catch((err) => {
  console.error('Fatal test runner error:', err);
  process.exit(1);
});
